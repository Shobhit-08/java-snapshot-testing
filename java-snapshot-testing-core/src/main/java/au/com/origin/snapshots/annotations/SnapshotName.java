package au.com.origin.snapshots.annotations;

import java.lang.annotation.*;

@Documented
@Target({ElementType.METHOD})
@Inherited
@Retention(RetentionPolicy.RUNTIME)
public @interface SnapshotName {
  String value();
}
