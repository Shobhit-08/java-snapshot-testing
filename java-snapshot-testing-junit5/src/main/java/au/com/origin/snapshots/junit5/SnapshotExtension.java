package au.com.origin.snapshots.junit5;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.SnapshotVerifier;
import au.com.origin.snapshots.config.PropertyResolvingSnapshotConfig;
import au.com.origin.snapshots.config.SnapshotConfig;
import au.com.origin.snapshots.config.SnapshotConfigInjector;
import au.com.origin.snapshots.exceptions.SnapshotMatchException;
import au.com.origin.snapshots.logging.LoggingHelper;

import java.lang.reflect.Field;
import java.util.Arrays;

import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.extension.*;
import org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor;
import org.junit.jupiter.engine.descriptor.ClassTestDescriptor;

@Slf4j
public class SnapshotExtension
        implements AfterAllCallback,
        BeforeAllCallback,
        SnapshotConfigInjector,
        ParameterResolver,
        BeforeEachCallback {

    private SnapshotVerifier snapshotVerifier;

    @Override
    public void beforeAll(ExtensionContext context) {
        // don't fail if a test is run alone from the IDE for example
        boolean failOnOrphans = shouldFailOnOrphans();
        Class<?> testClass =
                context
                        .getTestClass()
                        .orElseThrow(() -> new SnapshotMatchException("Unable to locate Test class"));
        this.snapshotVerifier = new SnapshotVerifier(getSnapshotConfig(), testClass, failOnOrphans);
    }

    @Override
    public void afterAll(ExtensionContext context) {
        this.snapshotVerifier.validateSnapshots();
    }

    @Override
    public SnapshotConfig getSnapshotConfig() {
        return new PropertyResolvingSnapshotConfig();
    }

    private boolean shouldFailOnOrphans() {
        return false;
    }

    @Override
    public boolean supportsParameter(
            ParameterContext parameterContext, ExtensionContext extensionContext)
            throws ParameterResolutionException {
        boolean supports = parameterContext.getParameter().getType() == Expect.class;
        if (supports) {
            LoggingHelper.deprecatedV5(
                    log,
                    "Injecting 'Expect' via method a argument is no longer recommended. Consider using instance variable injection instead.");
        }
        return supports;
    }

    @Override
    public Object resolveParameter(
            ParameterContext parameterContext, ExtensionContext extensionContext)
            throws ParameterResolutionException {
        return new Expect(
                snapshotVerifier,
                extensionContext
                        .getTestMethod()
                        .orElseThrow(() -> new RuntimeException("getTestMethod() is missing")));
    }

    @Override
    public void beforeEach(ExtensionContext context) {
        if (context.getTestInstance().isPresent() && context.getTestMethod().isPresent()) {
            Arrays.stream(context.getTestClass().get().getDeclaredFields())
                    .filter(it -> it.getType() == Expect.class)
                    .findFirst()
                    .ifPresent(
                            field -> {
                                Expect expect = Expect.of(snapshotVerifier, context.getTestMethod().get());
                                field.setAccessible(true);
                                try {
                                    field.set(context.getTestInstance().get(), expect);
                                } catch (IllegalAccessException e) {
                                    throw new RuntimeException(e);
                                }
                            });
        }
    }
}
