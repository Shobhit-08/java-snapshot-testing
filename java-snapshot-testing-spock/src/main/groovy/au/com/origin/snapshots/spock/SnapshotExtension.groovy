package au.com.origin.snapshots.spock

import au.com.origin.snapshots.SnapshotVerifier
import au.com.origin.snapshots.config.PropertyResolvingSnapshotConfig
import au.com.origin.snapshots.config.SnapshotConfig
import au.com.origin.snapshots.config.SnapshotConfigInjector
import org.spockframework.runtime.extension.AbstractAnnotationDrivenExtension
import org.spockframework.runtime.model.SpecInfo

class SnapshotExtension extends AbstractAnnotationDrivenExtension<EnableSnapshots> implements SnapshotConfigInjector {

    SnapshotVerifier snapshotVerifier;

    void visitSpecAnnotation(EnableSnapshots annotation, SpecInfo spec) {
        this.snapshotVerifier = new SnapshotVerifier(getSnapshotConfig(), spec.reflection, false)
    }

    void visitSpec(SpecInfo spec) {
        def snapshotMethodInterceptor = new SnapshotMethodInterceptor(snapshotVerifier)
        spec.allFeatures.featureMethod*.addInterceptor(snapshotMethodInterceptor)
        spec.addCleanupSpecInterceptor(snapshotMethodInterceptor)
    }

    @Override
    SnapshotConfig getSnapshotConfig() {
        return new PropertyResolvingSnapshotConfig()
    }
}