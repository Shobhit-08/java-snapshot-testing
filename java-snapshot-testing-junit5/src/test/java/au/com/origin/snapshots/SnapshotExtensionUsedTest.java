package au.com.origin.snapshots;

import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith(SnapshotExtension.class)
public class SnapshotExtensionUsedTest {

  private Expect expect;

  @Test
  public void shouldUseExtension(Expect expect) {
    expect.toMatchSnapshot("Hello World");
  }

  @Test
  public void shouldUseExtensionAgain(Expect expect) {
    expect.toMatchSnapshot("Hello World");
  }

  @Test
  public void shouldUseExtensionViaInstanceVariable() {
    expect.toMatchSnapshot("Hello World");
  }

  @Test
  public void shouldUseExtensionAgainViaInstanceVariable() {
    expect.toMatchSnapshot("Hello World");
  }

  @SnapshotName("hello_world")
  @Test
  public void snapshotWithName() {
    expect.toMatchSnapshot("Hello World");
  }

  @SnapshotName("hello_world_2")
  @Test
  public void snapshotWithNameAgain() {
    expect.toMatchSnapshot("Hello World");
  }
}
