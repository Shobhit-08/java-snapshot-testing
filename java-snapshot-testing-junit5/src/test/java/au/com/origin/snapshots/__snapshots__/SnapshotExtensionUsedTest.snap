au.com.origin.snapshots.SnapshotExtensionUsedTest.shouldUseExtension=[
Hello World
]


au.com.origin.snapshots.SnapshotExtensionUsedTest.shouldUseExtensionAgain=[
Hello World
]


au.com.origin.snapshots.SnapshotExtensionUsedTest.shouldUseExtensionAgainViaInstanceVariable=[
Hello World
]


au.com.origin.snapshots.SnapshotExtensionUsedTest.shouldUseExtensionViaInstanceVariable=[
Hello World
]


hello_world=[
Hello World
]


hello_world_2=[
Hello World
]